AX: [Site Isolation] Hit testing remote frames should account for the scroll offset

hoffmanjoshua opened the pull request:
#### 953140b3e76375ea875494bbdbd74019a8cec473
<pre>
AX: [Site Isolation] Hit testing remote frames should account for the scroll offset
<a href="https://bugs.webkit.org/show_bug.cgi?id=285700">https://bugs.webkit.org/show_bug.cgi?id=285700</a>
<a href="https://rdar.apple.com/142633437">rdar://142633437</a>

Reviewed by Chris Fleizach and Tyler Wilcock.

Prior to this patch, elements that were not in the visible bounds of a remote frame
would not be able to be hit tested when scrolled into view, since we did not take into
account the scroll position of the iframe. This now updates the page-level accessibility
hit test to account for this offset on both macOS and iOS.

* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.h:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase axObjectCache]):
(-[WKAccessibilityWebPageObjectBase remoteLocalFrame]):
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):

Canonical link: <a href="https://commits.webkit.org/288699@main">https://commits.webkit.org/288699@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c62714883ac46377aadb88eee735e04a2c2001f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84142 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3760 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38443 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89218 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35151 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11729 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65446 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23280 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87188 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2875 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76454 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45739 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2831 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34199 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73761 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31448 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90598 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11407 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8274 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73895 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11631 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72284 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73101 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18084 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17409 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2754 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11359 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16831 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11207 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14683 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12979 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -195,4 +193,18 @@ - (id)accessibilityFocusedUIElement
+        return nullptr;
+
+    auto page = m_page->corePage();
+    auto* remoteMainFrame = page ?  dynamicDowncast<WebCore::RemoteFrame>(page->mainFrame()) : nullptr;
```
Seems like there's an extra space after the `?`

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -195,4 +193,18 @@ - (id)accessibilityFocusedUIElement
+    if (!m_page)
+        return nullptr;
+
+    auto page = m_page->corePage();
```
Does `corePage()` return a `Page*`? If so, this should probably be `auto*` instead of `auto`.

hoffmanjoshua replied:
Ah yes, this was just moved from somewhere else—I'll fix it here

twilco commented on Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
```diff
@@ -74,8 +74,15 @@ - (id)accessibilityHitTest:(NSPoint)point
     WebCore::IntPoint convertedPoint = m_page->accessibilityScreenToRootView(WebCore::IntPoint(point));
+
+    // If we are hit-testing a remote element, offset the hit test by the scroll of the web page.
+    if (NakedPtr remoteLocalFrame = [self remoteLocalFrame]) {
```
Any reason we use a `NakedPtr` here, and as the return type of `remoteLocalFrame`? If it's feasible, I think a better option would be to return a raw `WebCore::LocalFrame*` from `remoteLocalFrame`, and turn that into a RefPtr here.

hoffmanjoshua replied:
I had originally done this to match `axObjectCache` in the web page object, since we use a NakedPtr for that. But, I don't see why we couldn't use a RefPtr instead. I'll update these

twilco commented on Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
```diff
@@ -74,8 +74,15 @@ - (id)accessibilityHitTest:(NSPoint)point
+
+    // If we are hit-testing a remote element, offset the hit test by the scroll of the web page.
+    if (NakedPtr remoteLocalFrame = [self remoteLocalFrame]) {
+        if (auto* frameView = remoteLocalFrame->view())
```
We call `scrollPosition()` on this `frameView`, which is not a [trivial](https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-calling-a-non-trivial-member-function-on-an-object-hold-a-smart-pointer-to-the-object-on-the-stack) function, so I think we have to use a smart pointer here. Looks like `CheckedPtr` is our only option for `FrameView`s.

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -62,10 +62,8 @@ @implementation WKAccessibilityWebPageObjectBase
-        auto& tree = remoteMainFrame->tree();
-        auto* firstFrame = dynamicDowncast<WebCore::LocalFrame>(tree.firstChild());
-        auto* document = firstFrame ? firstFrame->document() : nullptr;
+    } else if (auto remoteLocalFrame = self.remoteLocalFrame) {
```
Since `remoteLocalFrame` returns a pointer type, we should specify it instead of using `auto` (ideally we'd use a `RefPtr` if possible)

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
```diff
@@ -62,10 +62,8 @@ @implementation WKAccessibilityWebPageObjectBase
-        auto* firstFrame = dynamicDowncast<WebCore::LocalFrame>(tree.firstChild());
-        auto* document = firstFrame ? firstFrame->document() : nullptr;
+    } else if (auto remoteLocalFrame = self.remoteLocalFrame) {
+        auto* document = remoteLocalFrame ? remoteLocalFrame->document() : nullptr;
```
Since `axObjectCache()` is non-trivial, we should use a `CheckedPtr` to hold this document.

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
```diff
@@ -340,8 +340,13 @@ - (id)accessibilityHitTest:(NSPoint)point
-        if (auto* frameView = protectedSelf->m_page->localMainFrameView())
-            convertedPoint.moveBy(frameView->scrollPosition());
+        if (auto* localFrameView = protectedSelf->m_page->localMainFrameView()) {
+            fprintf(stderr, "\tlocalMainFrameView: %p\n", localFrameView);
```
Looks like this got leftover from debugging 🙂

hoffmanjoshua replied:
oops, removing!

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
```diff
@@ -340,8 +340,13 @@ - (id)accessibilityHitTest:(NSPoint)point
 
-        if (auto* frameView = protectedSelf->m_page->localMainFrameView())
-            convertedPoint.moveBy(frameView->scrollPosition());
+        if (auto* localFrameView = protectedSelf->m_page->localMainFrameView()) {
```
I think we need a smart pointer instead of a raw pointer here (I know you only changed the variable name, but good to fix while we're here)

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
```diff
@@ -340,8 +340,13 @@ - (id)accessibilityHitTest:(NSPoint)point
+        if (auto* localFrameView = protectedSelf->m_page->localMainFrameView()) {
+            fprintf(stderr, "\tlocalMainFrameView: %p\n", localFrameView);
+            convertedPoint.moveBy(localFrameView->scrollPosition());
+        } else if (NakedPtr remoteLocalFrame = [protectedSelf remoteLocalFrame]) {
```
Same comment about using `NakedPtr` over some type of smart pointer.

twilco commented on Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
```diff
@@ -340,8 +340,13 @@ - (id)accessibilityHitTest:(NSPoint)point
+            fprintf(stderr, "\tlocalMainFrameView: %p\n", localFrameView);
+            convertedPoint.moveBy(localFrameView->scrollPosition());
+        } else if (NakedPtr remoteLocalFrame = [protectedSelf remoteLocalFrame]) {
+            if (auto* frameView = remoteLocalFrame->view())
```
Another place where we want a smart pointer to hold this `FrameView`.

hoffmanjoshua committed:
AX: [Site Isolation] Hit testing remote frames should account for the scroll offset
https://bugs.webkit.org/show_bug.cgi?id=285700
rdar://142633437

Reviewed by Chris Fleizach and Tyler Wilcock.

Prior to this patch, elements that were not in the visible bounds of a remote frame
would not be able to be hit tested when scrolled into view, since we did not take into
account the scroll position of the iframe. This now updates the page-level accessibility
hit test to account for this offset on both macOS and iOS.

* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.h:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase axObjectCache]):
(-[WKAccessibilityWebPageObjectBase remoteLocalFrame]):
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):

Canonical link: https://commits.webkit.org/288699@main
